### ISSUE TITLE
Calibration of camera on Raspberry Pi 2 (slow pocessing)

### ISSUE THREAD

**Issue by onno182** [openalpr/openalpr#198]
openALPR works fine on Raspberry Pi 2, however average image recognition is between 4 - 8 seconds.  I need to get this down to below 2 seconds.

Image recognition seems to speed up substantially if the license plate occupies a greater part of the image.

What I missing is the ability to be able to tell openALPR to look within a particular area of the image (and not bother with the rest), in order to speed things up.  

Is there any functionality where I can set this without having to preprocess the image?.

**Comment by matthill** [comment]
How are you using OpenALPR, via the Daemon or the command line?

When I run on RPi2, I get much better performance than that... not sure why.  If I run all 4 cores simmultaneously, it can analyze around 5fps or so.

**Comment by favy87** [comment]
Hey , you can put an image of the SD raspberryPi that I also can perform tests?

**Comment by nglinh** [comment]
Hey Matt how do you use multi core on Rpi2? Did you compile all the dependencies ( the hard way in wiki)? I install the precompiled dependencies and compiled openalpr but it took like 10 seconds to analyze the sample image

**Comment by onno182** [comment]
I still havent had the time to check OpenALPR with the Daemon.  My results were with the binary.  Hopefully ill get a chance to do this this weekend.

I compiled all the dependencies by hand (the hard way).  Results vary very much by the image, anywhere from 2 seconds to about 10 or so for the larger image files.

Overclocking the Rpi2 substantially increases recognition time by about 2 seconds, which points to the processor.

A feature which might make ALPR work better embedded devices is to be able to tell it to scan a specified area in the image where the license plate is mostly likely to be located and perhaps providing the image in a form that makes recognition quickest.

**Comment by nglinh** [comment]
I think there Is already a detectregion method.im trying to detect the car first then limit openalpr to that region. But detecting the car itself is slow. I was able to get 3-4 fps for car detevtion using lbp features. Haar will be even slower. 


Matt mentioned something about openalpr on all 4 cores. I think that is a better approach





—

**Comment by devmad** [comment]
my test on a raspeberry Pi 2
7 seconds for one image
6 seconds before 'image not found" message
9 seconds for 3 images
so, 6 seconds for loading and only 1 second by image ?

with strace alpr i have a 5 seconds pause on a call mmap2 for 47Mo just after open runtime_data/ocr/tessdata/leu.trainddata (size=315188) ?

**Comment by nglinh** [comment]
What's your use case? More mine it's a video stream, so I used the openalpr integration library and just initialize it once. It's 7 seconds for initializing but it's a once time thing

**Comment by bosmart** [comment]
Anybody was able to make it work on all 4 cores? Mine is definitely using just 1 and it's slow. @matthill ?

**Comment by devmad** [comment]
The daemon openalprd with more than one camera is able to use one core by stream.
And i've read that in case of a multi-plates image, after the detection, the process can use multiple core ?

**Comment by awjreynolds** [comment]
Is openalpr restricted to single core?

**Comment by bosmart** [comment]
Also, what can be parallelized:
1. region detection (e.g. scale = thread)
2. recognition of multiple plates in a single image (region = thread)
3. recognition of one plate in an image (threshold = thread)

Personally I'm most interested in option 3 as it means always using multithreading.

**Comment by awjreynolds** [comment]
This may be the difference between the commercial versions of openALPR and
the one that is available for free. Turning off multi-core recognition will
significantly reduce performance. http://www.openalpr.com/pricing.html

If this is the case, then it's a real shame as Arduino boards focus on
providing 4-16 low power cores.

**Comment by Zapnologica** [comment]
@awjreynolds  since when do arduino board have 4 - 16 cores?

**Comment by awjreynolds** [comment]
https://www.raspberrypi.org/products/raspberry-pi-2-model-b/


   - A 900MHz quad-core ARM Cortex-A7 CPU

**Comment by marcosnils** [comment]
> When I run on RPi2, I get much better performance than that... not sure why. If I run all 4 cores simmultaneously, it can analyze around 5fps or so.

@matthill  Any hints on how to make this work?

**Comment by matthill** [comment]
One quick way to test multi-core OpenALPR is to run the commercial code connecting to the cloud.  There's an ARM version compiled for Ubuntu 14.04 in beta.  You can run install it via:
curl <(http://deb.openalpr.com/install-beta)

Then set the analysis_threads to 4 in /etc/openalpr/alprd.conf and send the data up to the cloud.  The log will generally show you how many recognitions it's getting, and you should see all 4 cores in use when there is activity on the video stream.

You could also use OpenALPR library to analyze multiple images in parallel to get your benchmark.  

I think the reason why you're seeing such high delay is the initialization.  When you run the OpenALPR binary, it pulls in all the training data -- on RPi that takes many seconds.  But it only has to happen once at startup.  With the library, the library (or daemon) can be used again and again and not reinitialize.  When you run the alpr CLI program, it reinitializes each time.

**Comment by newsages** [comment]
GPU support delay and make false positive?
Try compile w/o gpu support.

**Comment by marktheunissen** [comment]
@newsages How much of a difference does that make?

**Comment by nunonup** [comment]
I have OpenALPR running in a python script that initializes OpenALPR on the start and stays 'on'! When a picture is send to 'decode', the result is almost instantaneous, less than 1sec... RPi3.

**Comment by marktheunissen** [comment]
@nunonup Nice! I'm curious, how big is the image? and did you compile OpenALPR specifically for Raspi or tweak any other settings?

**Comment by nunonup** [comment]
Yes, i compile OpenALPR specifically for raspi. Then i run a script that starts OpenALPR and i have a sensor that detects a car. When a car is detected a foto is taken and decoded by OpenALPR. Is quick because OpenALPR doesn't need to start, is already running...
I spend some time to put it working fine... :)

**Comment by vladimirr0708982** [comment]
what sensor u use ? inductive or some pirr sensor ?

**Comment by marktheunissen** [comment]
How big is the image? Resolution?

**Comment by nunonup** [comment]
Any kind of sensor... I use GPIO... When goes up, active a usb cam and take a foto... 640x480. More than that doesn't increase the preacision and takes a bit more...
